### ISSUE TITLE
[ROOT-9733] TRandom3 does not implement perfectly a Mersenne Twister PRNG

### ISSUE THREAD

**Issue by vepadulano** [root-project/root#14581]
### Check duplicate issues.

- [ ] Checked for duplicates

### Description

From https://its.cern.ch/jira/browse/ROOT-9733

The {{TRandom3}} is a random number engine based on Mersenne Twister algorithm, defined in {{TRandom3.cxx}}. Since 2002 (see [commit 026ec9c0f458be61825e7335ec51e115ec6cbf4c](https://github.com/root-project/root/commit/026ec9c0f458be61825e7335ec51e115ec6cbf4c)), an eventual zero return value is discarded by {{TRandom3::Rndm()}}. This does not allow the algorithm to return the same sequence of the C++11 standard implementation {{std::mt19937}} (when seeded with the same value - except for seed zero, since handled in a special way by {{TRandom3::SetSeed()}} - and multiplied by 2^-32).

A related mistake is present also in the description of the class, where it states the algorithm period to be 2^19937 - 1. If you discard the zeros, of course the period will be (very slightly) less than it.

In conclusion, {{TRandom3}} is an imperfect implementation of the Mersenne Twister 32-bit algoritm. Furthermore, the documentation does not explain this small difference.

 

### Reproducer

As per the description

### ROOT version

Any

### Installation method

Any

### Operating system

Any

**Comment by ferdymercury** [comment]
Another issue is that the docu says it includes 1, but it doesn't, since:

`UINT_MAX* 2.3283064365386963e-10`
is
`0.999999999767169356`

as noted in https://github.com/root-project/root/commit/07c0ad7ef9098fb58e800903d203c472c5ec5b1f
